### ISSUE TITLE
help: Trouble retrieving user details from OIDC

### ISSUE THREAD

**Issue by icsy7867** [uxbox/uxbox#3130]
### Steps To Reproduce

Using docker-compose and my companies OIDC configuration, try to log into penpot using OIDC.

### Expected behavior

To successfully log in

### Actual behavior

You are taken back to the main login screen with error codes in the URL.
https://penpot.mycompany.co/#/auth/login?error=unable-to-auth&hint=unable+to+retrieve+user+info

### Desktop (please complete the following information)

MacOs 13.3
Chrome Version 108.0.5359.98 (Official Build) (arm64)

### Environment (please complete the following information)

Using podman and docker-compose

### Backend Stack Trace

```console
+ exec /opt/jdk/bin/java -Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager -Dlog4j2.configurationFile=log4j2.xml -XX:-OmitStackTraceInFastThrow -jar penpot.jar -m app.main
2023-04-12 15:42:15,004 main INFO Registered Log4j as the java.util.logging.LogManager.
[2023-04-12 15:42:27.023] I app.auth.oidc - hint="provider initialized", provider="oidc", method="discover", client-id="02753b18-ced1-443e-8a5c-364c2c7355d9", client-secret="DCKFc***********************************", scopes="profile,email,openid", auth-uri="https://penpot.mycompany.co/adfs/oauth2/authorize/", user-uri="https://penpot.mycompany.co/adfs/userinfo", token-uri="https://penpot.mycompany.co/adfs/oauth2/token/", roles-attr=nil, roles=nil
[2023-04-12 15:42:27.026] I app.metrics - action="initialize metrics"
[2023-04-12 15:42:27.042] I app.db - hint="initialize connection pool", name="main", uri="postgresql://penpot-postgres/penpot", read-only=false, with-credentials=true, min-size=0, max-size=60
[2023-04-12 15:42:27.067] I app.migrations - hint="running migrations", module=:app.migrations/migrations
[2023-04-12 15:42:27.307] W app.setup - hint="using autogenerated secret-key, it will change on each restart and will invalidate all sessions on each restart, it is hightly recommended setting up the PENPOT_SECRET_KEY environment variable"
[2023-04-12 15:42:27.328] I app.redis - hint="initialize redis resources", uri="redis://penpot-redis/0", io-threads=8, worker-threads=8, connect?=nil
[2023-04-12 15:42:27.454] I app.msgbus - hint="initialize msgbus", buffer-size=128
[2023-04-12 15:42:27.799] I app.setup.builtin-templates - hint="loading template files", total=9
[2023-04-12 15:42:28.652] I app.http - hint="starting http server", port=6060, host="0.0.0.0"
[2023-04-12 15:42:28.673] I io.undertow - starting server: Undertow - 2.3.0.Final
[2023-04-12 15:42:28.682] I org.xnio - XNIO version 3.8.8.Final
[2023-04-12 15:42:28.687] I org.xnio.nio - XNIO NIO Implementation Version 3.8.8.Final
[2023-04-12 15:42:28.704] I org.jboss.threads - JBoss Threads version 3.5.0.Final
[2023-04-12 15:42:28.748] I app.loggers.database - hint="initializing database error persistence"
[2023-04-12 15:42:28.750] I app.storage.tmp - hint="started tmp file cleaner"
[2023-04-12 15:42:28.755] I app.worker - hint="registry initialized", tasks=13
[2023-04-12 15:42:28.758] I app.worker - hint="cron: started", tasks=8
[2023-04-12 15:42:28.780] I app.worker - hint="dispatcher: started"
[2023-04-12 15:42:28.782] I app.worker - hint="monitor: started", name="default"
[2023-04-12 15:42:28.783] I app.worker - hint="worker: started", worker-id=0, queue="default"
[2023-04-12 15:42:28.784] I app.worker - hint="worker: started", worker-id=0, queue="webhooks"
[2023-04-12 15:42:28.785] I app.main - hint="welcome to penpot", flags="login-with-password,backend-api-doc,secure-session-cookies,log-emails,backend-worker,login-with-oidc,smtp,email-verification,registration", worker?=true, version="1.18.1-9810-g6d3baaae4"
[2023-04-12 15:42:46.075] E app.auth.oidc - hint="error on oauth process"
SUMMARY:
 →  clojure.lang.ExceptionInfo: unable to retrieve user info (oidc.clj:324)
DETAIL:
 →  clojure.lang.ExceptionInfo: unable to retrieve user info (oidc.clj:324)
    at: app.auth.oidc$retrieve_user_info$validate_response__34244.invoke(oidc.clj:324)
        promesa.util.Function.apply(util.cljc:40)
        java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
        java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
        java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
        java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
        java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
    dt: {:type :internal,
         :code :unable-to-retrieve-user-info,
         :hint "unable to retrieve user info",
         :http-status 401,
         :http-body ""}
```
```


### Additional context

So the OIDC process seems to work, I log in with my username password, then we get our duo 2 factor, this all fires off without a hitch.  However I am then brought back to the main login with the URL mentioned before:
https://penpot.mycompany.co/#/auth/login?error=unable-to-auth&hint=unable+to+retrieve+user+info

I verified with IT, that the claims/props being returned are "name" and "email"

Also my config for this is pretty simple, and for testing I set the name and email attributes just in case.
```
PENPOT_OIDC_BASE_URI=https://oidcsso.mycompany.co/adfs/oauth2
PENPOT_OIDC_CLIENT_ID=XXXXXXXXXX
PENPOT_OIDC_CLIENT_SECRET=YYYYYYYYY
# Attribute to use for lookup the name on the user object. Optional,
# if not perovided, the `name` prop will be used.
PENPOT_OIDC_NAME_ATTR=name
PENPOT_OIDC_EMAIL_ATTR=email
```

We have:
https://oidcsso.mycompany.co/adfs/.well-known/openid-configuration
And all of the values here match what penpot detects for the URLS.

I can't seem to find any more helpful information.  The OIDC/SSO part seems to be working fine as far as IT is telling me, and I am at a loss of what else to check.

**Comment by icsy7867** [comment]
After reading as much as I can on OIDC and then different ways of using it, I have found out that Microsoft's ADFS implementation of OIDC does not follow the same logic that other OIDC do.

For penpot to support OIDC via ADFS, the process is a little bit different.  Normally, if I understand it correctly the OIDC path would be to:
Authorize -> Token -> Userinfo
However with ADFS, The userinfo API endpoint ONLY contains "sub". The OpenID spec only requires the SUB claim here.  All of the other claims are returned in the Token API.

Additionally when requesting the initial authorization code, you are also supposed to send:
resource=urn:microsoft:userinfo

But utlimately when using ADFS, all of the claims are provided in the JWT Token.  I guess from Microsofts POV this saves a step.  Here is a similar post that was made for portainer:
https://github.com/portainer/portainer/issues/7044
However everyone keeps assuming that Microsoft Azure OAUTH = AD FS OAUTH which is unfortunately not the case.

https://serverfault.com/questions/1109792/setting-up-oidc-with-adfs-invalid-userinfo-request
https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/development/ad-fs-openid-connect-oauth-concepts

```
https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/overview/ad-fs-faq#i-am-trying-to-get-additional-claims-on-the-user-info-endpoint-but-its-only-returning-subject-how-can-i-get-additional-claims

I'm trying to get more claims on the UserInfo endpoint, but it's only returning subject. How can I get more claims?
The AD FS UserInfo endpoint always returns the subject claim as specified in the OpenID standards. AD FS doesn't support additional claims requested via the UserInfo endpoint. If you need more claims in an ID token, see [Custom ID tokens in AD FS](https://learn.microsoft.com/en-us/windows-server/identity/ad-fs/development/custom-id-tokens-in-ad-fs).

```
**Here is another product that implemented/resovled this by giving a method of using the ID_Token instead of the /userinfo endpoint:**
https://github.com/mozilla/mozilla-django-oidc/blob/main/mozilla_django_oidc/auth.py#L291
OIDC_STORE_ID_TOKEN

Possible Fixes -> 
Add either a Microsoft ADFS or give the ability to get the claims from the ID_Token.
OR
Implement something like header authentication support: https://github.com/penpot/penpot/issues/1766


Header authentication is also great because you can use a reverse proxy like Mellon, or something like Authelia, and it can work with all sorts of things (basic auth, SAML, oidc, etc...) as long as there is a proxy or something to support it.
Here are a couple of examples of common tools that use header auth:
https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/auth-proxy/
https://demo.netbox.dev/static/docs/administration/authentication/

**Also... looking at the original OIDC request for penpot**... why even call userinfo anyways?  From where you guys were going back and forth, after showing the authorization, ID_Token and userinfo values, why not just stop at ID_Token since it has all the required values as well?
https://github.com/penpot/penpot/issues/810

ID_Token
```
{
  "exp": 1637219408,
  "iat": 1637219108,
  "auth_time": 0,
  "jti": "004f1aac-1dc7-42d3-a5e8-e85ea289e198",
  "iss": "https://<server>/auth/realms/<realm>",
  "aud": "<client>",
  "sub": "187dd782-beb9-438d-bde4-2b5a146693d8",
  "typ": "ID",
  "azp": "<client>",
  "session_state": "11321e37-590d-4ac3-bc1c-b16b14ef931d",
  "acr": "1",
  "sid": "33321e39-590d-4ac3-bc1c-b16b14ef931d",
  "email_verified": true,
  "roles": [
    "admin",
    "manage-account",
    "manage-account-links",
    "view-profile"
  ],
  "name": "<fullname>",
  "preferred_username": "<email>",
  "locale": "de",
  "given_name": "<firstname>",
  "family_name": "<lastname>",
  "email": "<email>"
}
```

~~As a quick and dirty test:
https://github.com/penpot/penpot/blob/afb09919edf056744e3eab71d997b3ec072e4692/backend/src/app/auth/oidc.clj~~
Just kidding.  I thought I could take a stab at editing the clojure app, and rebuilding the containers, however I need to learn more about clojure before I can do anything useful.
```
      (= 200 (:status response))
      (let [data      (json/decode (:body response))
            token-uri (get data :token_endpoint)
            auth-uri  (get data :authorization_endpoint)
            user-uri  (get data :userinfo_endpoint)]
        (l/debug :hint "oidc uris discovered"
                 :token-uri token-uri
                 :auth-uri auth-uri
                 :user-uri user-uri)
        {:token-uri token-uri
         :auth-uri  auth-uri
         :user-uri  user-uri})
```
~~It seems like you could just use the same data from the token URI instead of going to the userinfo:~~
```
~~user-uri  (get data :token_endpoint)]~~
```

**Comment by niwinz** [comment]
Should be fixed with https://github.com/penpot/penpot/issues/3152

**Comment by icsy7867** [comment]
> Should be fixed with #3152

With my organization, we use ADFS, but our parent org uses CAS, which uses a more traditional OpenID and they handle the attributes a bit weird, and your addition to allow for nested prop values is a welcome fix.

However, with ADFS and openid (Similar to Azure AD I believe), they do not include the claims/props in the /userinfo API method.  They are included in the token.  So I dont think this would be fixed for that specific purpose.  For ADFS (And possibly others), there would need to be modifications and/or a flag to get the attributes from the token and not the userinfo endpoint.

I was looking at the commit changes, and I am not familiar enough with clojure to tell if this was changed to search the token for the attributes as well as /userinf, but just wanted to clarify.

Also I tried to snag the development branch and do a buld-bundle and a build-docker-images to test.  everything seemed to complete, but when I tried to use those images, the backend would crash with no info in the docker logs for the container.

In any case thank you for your hard work and I am looking forward to the fixes/improvements!
